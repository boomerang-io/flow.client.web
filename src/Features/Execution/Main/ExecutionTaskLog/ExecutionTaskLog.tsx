import React, { useState } from "react";
import { SkeletonPlaceholder } from "@carbon/react";
import { TooltipHover } from "@boomerang-io/carbon-addons-boomerang-react";
import { UseQueryResult } from "react-query";
import TaskItem from "./TaskItem";
import orderBy from "lodash/orderBy";
import { getSimplifiedDuration } from "Utils/timeHelper";
import { QueryStatus } from "Constants";
import { executionStatusIcon, ExecutionStatusCopy } from "Constants";
import { WorkflowExecution } from "Types";
import { ArrowsVertical, ChevronLeft } from "@carbon/react/icons";
import styles from "./executionTaskLog.module.scss";

type Props = {
  workflowExecution: UseQueryResult<WorkflowExecution, Error>;
};

function ExecutionTaskLog({ workflowExecution }: Props) {
  const [isCollapsed, setIsCollapsed] = useState(false);
  const [tasksSort, setTasksSort] = useState<boolean | "desc" | "asc">("desc");

  const toggleCollapse = () => {
    setIsCollapsed(!isCollapsed);
  };

  const toggleSort = () => {
    setTasksSort(tasksSort === "desc" ? "asc" : "desc");
  };

  if (workflowExecution.status !== QueryStatus.Success) {
    return (
      <aside className={`${styles.container} ${isCollapsed ? styles.collapsed : ""}`}>
        <section className={styles.statusBlock}>
          <SkeletonPlaceholder className={styles.statusBlockSkeleton} />
        </section>
        <section className={styles.taskbar}>
          <p className={styles.taskbarTitle}>Task log</p>
          {!isCollapsed && (
            <TooltipIco
              disabled
              align="center"
              className={styles.taskbarButton}
              id="sort-tooltip"
              data-testid="taskbar-button"
            >
              <ArrowsVertical size={32} className={styles.taskbarArrows} />
            </TooltipIco>
          )}
        </section>
        <ul className={styles.tasklog}>
          <SkeletonPlaceholder className={styles.taskLogSkeleton} />
        </ul>
      </aside>
    );
  }

  const { id, duration, status, steps } = workflowExecution.data;
  const Icon = executionStatusIcon[status];
  const sortedTasks = steps ? orderBy(steps, (step: any) => step.order, [tasksSort]) : [];

  return (
    <aside className={`${styles.container} ${isCollapsed ? styles.collapsed : ""}`}>
      <section className={`${styles.statusBlock} ${styles[status]}`}>
        <div className={styles.duration}>
          <p className={styles.title}>Duration</p>
          <time className={styles.value}>
            {typeof duration === "number" ? getSimplifiedDuration(duration / 1000) : "--"}
          </time>
        </div>
        <div className={styles.status}>
          <p className={styles.title}>Status</p>
          <div className={styles.statusData}>
            {Icon && <Icon aria-label={status} className={styles.statusIcon} />}
            <p className={styles.value}>{status ? ExecutionStatusCopy[status] : "--"}</p>
          </div>
        </div>
        <button className={styles.collapseButton} onClick={toggleCollapse}>
          <ChevronLeft size={32} className={styles.chevron} />
        </button>
      </section>
      <section className={styles.taskbar}>
        <p className={styles.taskbarTitle}>Task log</p>
        {!isCollapsed && (
          <TooltipHover
            align="center"
            className={styles.taskbarButton}
            id="sort-tooltip"
            data-testid="taskbar-button"
            direction="top"
            onClick={toggleSort}
            tooltipText="Change sort direction (by start time)"
          >
            <ArrowsVertical size={32} className={styles.taskbarArrows} />
          </TooltipHover>
        )}
      </section>
      <ul className={styles.tasklog}>
        {sortedTasks.map((step) => (
          <TaskItem
            key={step.id}
            flowActivityId={id}
            hidden={isCollapsed}
            task={step}
            executionId={workflowExecution.data.id}
          />
        ))}
      </ul>
    </aside>
  );
}

export default ExecutionTaskLog;
