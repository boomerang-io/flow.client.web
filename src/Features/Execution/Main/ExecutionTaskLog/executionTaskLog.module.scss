@import "Styles/animation";
@import "Styles/mixins";

.container {
  background-color: var(--cds-ui-01);
  transition: transform $carbon-motion--timing--productive $carbon-motion--easing--productive;
  width: 24.5rem;

  & + div {
    transition: margin-left $carbon-motion--timing--productive $carbon-motion--easing--productive;
  }

  &.collapsed {
    transform: translate3d(-22.5rem, 0, 0);

    .duration,
    .status,
    .taskbar,
    .tasklog {
      opacity: 0.01;
    }

    .chevron {
      transform: rotate(180deg);
    }

    & + div {
      margin-left: -22.5rem;
    }
  }
}

.statusBlock {
  color: var(--cds-text-01);
  display: flex;
  height: 7rem;
  overflow: hidden;
  padding: 1.5rem 0 1.5rem 1.5rem;
  position: relative;
  width: 100%;

  &.completed {
    background-color: #92eeee;
  }

  &.failure, &.cancelled {
    background-color: #fcd0d3;
  }

  &.inProgress,
  &.notstarted,
  &.invalid,
  &.skipped {
    background-color: #d5d9e0;
  }
}

.duration {
  margin-right: 2.5rem;
  opacity: 0.99;
  transition: opacity $carbon-motion--timing--productive $carbon-motion--easing--productive;
}

.title {
  font-size: 0.875rem;
  line-height: 1.125rem;
  margin-bottom: 0.25rem;
}

.value {
  font-size: 2rem;
  font-weight: 600;
  line-height: 2.625rem;
}

.status {
  opacity: 0.99;
  transition: opacity $carbon-motion--timing--productive $carbon-motion--easing--productive;
}

.statusData {
  align-items: center;
  display: flex;
  white-space: nowrap;
}

.statusIcon {
  height: 2rem;
  margin-right: 0.5rem;
  width: 2rem;

  path {
    fill: var(--cds-text-01);
  }
}

.collapseButton {
  background-color: transparent;
  border: none;
  cursor: pointer;
  line-height: 0rem;
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;

  &:focus {
    @include focus-outline();
  }
}

.chevron {
  height: 1rem;
  transition: transform $carbon-motion--timing--productive $carbon-motion--easing--productive;
  width: 1rem;

  path {
    fill: var(--cds-text-01);
  }
}

.statusBlockSkeleton {
  width: 90%;
}

.taskbar {
  align-items: center;
  background-color: transparent;
  display: flex;
  height: 3.25rem;
  justify-content: space-between;
  opacity: 0.99;
  padding: 1rem 0.875rem 1rem 1.5rem;
  transition: opacity $carbon-motion--timing--productive $carbon-motion--easing--productive;
  width: 100%;
}

.taskbarTitle {
  color: var(--cds-text-01);
  font-size: 1rem;
  line-height: 1.25rem;
  white-space: nowrap;
}

.taskbarButton {
  background-color: transparent;
  border: none;
  cursor: pointer;
  height: 1.25rem;
  transition: transform 0.2s ease;
  width: 1.25rem;

  &:focus {
    @include focus-outline();
  }
}

.taskbarArrows {
  height: 1.25rem;
  width: 1.25rem;
}

.tasklog {
  height: calc(100% - 10.25rem);
  overflow-y: auto;
  opacity: 0.99;
  padding-bottom: 2rem;
  scroll-behavior: smooth;
  transition: opacity $carbon-motion--timing--productive $carbon-motion--easing--productive;
  width: 100%;
}

.taskLogSkeleton {
  height: 12.5rem;
  margin-left: 1.5rem;
  width: 20.75rem;
}
