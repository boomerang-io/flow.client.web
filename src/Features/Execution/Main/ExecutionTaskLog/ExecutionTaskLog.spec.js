import React from "react";
import { fireEvent } from "@testing-library/react";
import ExecutionTaskLog from "./index";

const props = {
  workflowExecution: {
    status: "success",
    data: {
      id: "1",
      duration: 904934,
      status: "completed",
      steps: [
        {
          activityId: "5c36289096052900012cc81d",
          duration: 300190,
          flowTaskStatus: "completed",
          id: "5c36289096052900012cc81e",
          order: 1,
          startTime: "2019-09-03T15:00:00.230+0000",
          taskId: "d53a65a5-be1b-4e58-b0e5-0173c5aabe47",
          outputs: {
            args: {
              test: "true",
            },
            data: "",
            files: {},
            form: {},
            headers: {
              Accept: "/",
              "Accept-Encoding": "gzip,deflate",
              Connection: "close",
              "Content-Type": "text/http",
              Host: "httpbin.org",
              "User-Agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
            },
            json: null,
            method: "GET",
            origin: "147.10.60.191",
            url: "https://httpbin.org/anything?test=true",
          },
          taskName: "Slack",
        },
        {
          activityId: "5c36289096052900012cc81s",
          duration: 300190,
          flowTaskStatus: "inProgress",
          id: "5c3628909605290001dcc81e",
          order: 2,
          startTime: "2019-09-03T15:01:00.103+0000",
          taskId: "ced6eda1-47b0-4b8a-bddc-1c0bf64e0b84",
          outputs: {
            args: {
              test: "true",
            },
            data: "",
            files: {},
            form: {},
            headers: {
              Accept: "/",
              "Accept-Encoding": "gzip,deflate",
              Connection: "close",
              "Content-Type": "text/http",
              Host: "httpbin.org",
              "User-Agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
            },
            json: null,
            method: "GET",
            origin: "147.10.60.191",
            url: "https://httpbin.org/anything?test=true",
          },
          taskName: "Email",
        },
      ],
    },
  },
};

describe("ExecutionTaskLog --- Snapshot", () => {
  it("Capturing Snapshot of ExecutionTaskLog", () => {
    const { baseElement } = rtlRouterRender(<ExecutionTaskLog {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("ExecutionTaskLog --- RTL", () => {
  it("Sort tasks", () => {
    const { getByTestId, getAllByTestId } = rtlRouterRender(<ExecutionTaskLog {...props} />);

    const sortButton = getByTestId("taskbar-button");
    let taskItems = getAllByTestId("taskitem-name");
    expect(taskItems[0]).toHaveTextContent("Email");
    expect(taskItems[1]).toHaveTextContent("Slack");

    fireEvent.click(sortButton);
    taskItems = getAllByTestId("taskitem-name");
    expect(taskItems[0]).toHaveTextContent("Slack");
    expect(taskItems[1]).toHaveTextContent("Email");
  });
});
