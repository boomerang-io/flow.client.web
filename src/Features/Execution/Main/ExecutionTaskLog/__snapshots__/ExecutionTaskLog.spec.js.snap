// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecutionTaskLog --- Snapshot Capturing Snapshot of ExecutionTaskLog 1`] = `
<body
  id="app"
>
  <div>
    <aside
      class="container "
    >
      <section
        class="statusBlock completed"
      >
        <div
          class="duration"
        >
          <p
            class="title"
          >
            Duration
          </p>
          <time
            class="value"
          >
            15min
          </time>
        </div>
        <div
          class="status"
        >
          <p
            class="title"
          >
            Status
          </p>
          <div
            class="statusData"
          >
            <svg
              aria-label="completed"
              class="statusIcon"
              fill="currentColor"
              focusable="false"
              height="32"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 32 32"
              width="32"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
              />
              <path
                d="M14 21.5L9 16.54 10.59 14.97 14 18.35 21.41 11 23 12.58 14 21.5z"
              />
            </svg>
            <p
              class="value"
            >
              Succeeded
            </p>
          </div>
        </div>
        <button
          class="collapseButton"
        >
          <svg
            aria-hidden="true"
            class="chevron"
            fill="currentColor"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 16L20 6 21.4 7.4 12.8 16 21.4 24.6 20 26z"
            />
          </svg>
        </button>
      </section>
      <section
        class="taskbar"
      >
        <p
          class="taskbarTitle"
        >
          Task log
        </p>
        <button
          aria-describedby="sort-tooltip"
          class="bx--tooltip__trigger bx--tooltip--a11y taskbarButton bx--tooltip--top bx--tooltip--align-center"
          data-testid="taskbar-button"
          type="button"
        >
          <span
            class="bx--assistive-text"
            id="sort-tooltip"
          >
            Change sort direction (by start time)
          </span>
          <svg
            aria-hidden="true"
            class="taskbarArrows"
            fill="currentColor"
            focusable="false"
            height="32"
            preserveAspectRatio="xMidYMid meet"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
            />
          </svg>
        </button>
      </section>
      <ul
        class="tasklog"
      >
        <li
          class="taskitem inProgress"
          id="task-ced6eda1-47b0-4b8a-bddc-1c0bf64e0b84"
          tabindex="0"
        >
          <div
            class="progressBar"
          />
          <section
            class="header"
          >
            <div
              class="title"
            >
              <svg
                aria-label="inProgress"
                class="taskIcon"
                fill="currentColor"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15 11H17V20H15zM13 2H19V4H13z"
                />
                <path
                  d="M28,9,26.58,7.59,24.33,9.84a10.94,10.94,0,1,0,1.18,1.65ZM16,26a9,9,0,1,1,9-9A9,9,0,0,1,16,26Z"
                />
              </svg>
              <p
                data-testid="taskitem-name"
                title="Email"
              >
                Email
              </p>
            </div>
            <div
              class="status inProgress"
            >
              <svg
                aria-label="inProgress"
                class="statusIcon"
                fill="currentColor"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M15 11H17V20H15zM13 2H19V4H13z"
                />
                <path
                  d="M28,9,26.58,7.59,24.33,9.84a10.94,10.94,0,1,0,1.18,1.65ZM16,26a9,9,0,1,1,9-9A9,9,0,0,1,16,26Z"
                />
              </svg>
              <p>
                In Progress
              </p>
            </div>
          </section>
          <section
            class="data"
          >
            <div
              class="time"
            >
              <p
                class="timeTitle"
              >
                Start time
              </p>
              <time
                class="timeValue"
              >
                12:00:00 AM
              </time>
            </div>
            <div
              class="time"
            >
              <p
                class="timeTitle"
              >
                Duration
              </p>
              <time
                class="timeValue"
              >
                5 mins
              </time>
            </div>
          </section>
          <section
            class="data"
          />
        </li>
        <li
          class="taskitem completed"
          id="task-d53a65a5-be1b-4e58-b0e5-0173c5aabe47"
          tabindex="0"
        >
          <div
            class="progressBar"
          />
          <section
            class="header"
          >
            <div
              class="title"
            >
              <svg
                aria-label="completed"
                class="taskIcon"
                fill="currentColor"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
                />
                <path
                  d="M14 21.5L9 16.54 10.59 14.97 14 18.35 21.41 11 23 12.58 14 21.5z"
                />
              </svg>
              <p
                data-testid="taskitem-name"
                title="Slack"
              >
                Slack
              </p>
            </div>
            <div
              class="status completed"
            >
              <svg
                aria-label="completed"
                class="statusIcon"
                fill="currentColor"
                focusable="false"
                height="32"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
                />
                <path
                  d="M14 21.5L9 16.54 10.59 14.97 14 18.35 21.41 11 23 12.58 14 21.5z"
                />
              </svg>
              <p>
                Succeeded
              </p>
            </div>
          </section>
          <section
            class="data"
          >
            <div
              class="time"
            >
              <p
                class="timeTitle"
              >
                Start time
              </p>
              <time
                class="timeValue"
              >
                12:00:00 AM
              </time>
            </div>
            <div
              class="time"
            >
              <p
                class="timeTitle"
              >
                Duration
              </p>
              <time
                class="timeValue"
              >
                5 mins
              </time>
            </div>
          </section>
          <section
            class="data"
          />
        </li>
      </ul>
    </aside>
  </div>
</body>
`;
